### PR TITLE
Enhance Epoch information for Pulse train epoch

### LINKED ISSUE #805: Epoch information is lacking pulse high times
The epoch info of a stimset with 5 pulse train pulses looks like

```
0.0000000,0.0200000,Inserted TP;Test Pulse;,0,
0.0000000,0.0050000,Baseline,1,
0.0050000,0.0150000,Inserted TP;Test Pulse;pulse;Amplitude=10;,1,
0.0150000,0.0200000,Baseline,1,
0.0200000,0.4300050,Stimset,0,:0.0200000,0.4300050,Epoch=0;Type=Pulse Train;Amplitude=1;,1,
0.0200000,0.1200000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=0;,2,
0.1200000,0.2200000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=1;,2,
0.2200000,0.3200000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=2;,2,
0.3200000,0.4200000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=3;,2,
0.4200000,0.4300050,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=4;,2,
```

And if we just look at the first pulse

```
0.0200000,0.1200000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=0;,2,:
```

grafik

this reveals that we don't have the information about the pulse baseline as separate info. But we need that for the enhanced spike control.

So we need new entries like

```
0.0200000,0.0300000,Epoch=0;Type=Pulse Train;Amplitude=1;Pulse=0;Active,3,:
0.0300000,0.1200000,Epoch=0;Type=Pulse Train;Pulse=0;Baseline,3,:
```

### PR DESCRIPTION
**Pull request by t-b**
First preparational branch for MSQ_SpikeControl enhancements.

- [x] CI pass
- [x] Let MH have a look at the epoch changes

Close #805

### CODE REVIEW

**Comment by t-b** [comment]
@MichaelHuth Sorry I had to fix the NI test case failure. The reason is that it had a single point baseline at the end, and the new checks in TestEpochsMonotony barfed at that. I'm now not checking those. I've made the following changes compared to 0ed9a210 (PA plot: Gather more pulse infos, 2021-02-02).

```diff
diff --git a/Packages/Testing-MIES/UTF_Epochs.ipf b/Packages/Testing-MIES/UTF_Epochs.ipf
index 30ee804ed..972c0ea36 100644
--- a/Packages/Testing-MIES/UTF_Epochs.ipf
+++ b/Packages/Testing-MIES/UTF_Epochs.ipf
@@ -3,8 +3,6 @@
 #pragma rtFunctionErrors=1
 #pragma ModuleName=Epochs
 
-Constant EPOCHS_RESERVED_RANGE_MS = 0.05
-
 /// @brief Acquire data with the given DAQSettings on two headstages
 static Function AcquireData(s, devices, stimSetName1, stimSetName2[, dDAQ, oodDAQ])
 	STRUCT DAQSettings& s
@@ -110,13 +108,14 @@ static Function TestEpochChannelTight(e)
 End
 
 static Constant OODDAQ_PRECISION = 0.001
+static Constant OTHER_EPOCHS_PRECISION = 0.05
 
 static Function TestEpochsMonotony(e, DAChannel)
 	WAVE/T e
 	WAVE DAChannel
 
 	variable i, epochCnt, rowCnt, beginInt, endInt, epochNr, dur, amplitude, center, DAAmp
-	variable level
+	variable first, last, level, range
 	string s, name
 
 	rowCnt = DimSize(e, ROWS)
@@ -198,6 +197,9 @@ static Function TestEpochsMonotony(e, DAChannel)
 	for(i = 0; i < epochCnt; i += 1)
 		name  = e[i][2]
 		level = str2num(e[i][3])
+		first = startT[i] * 1000
+		last  = endT[i] * 1000
+		range = last - first
 
 		// check amplitudes
 		if(strsearch(name, "Amplitude", 0) > 0)
@@ -205,23 +207,24 @@ static Function TestEpochsMonotony(e, DAChannel)
 			amplitude = NumberByKey("Amplitude", name, "=")
 			CHECK(IsFinite(amplitude))
 
-			WaveStats/R=(startT[i] * 1000, endT[i] * 1000)/Q/M=1 DAChannel
+			WaveStats/R=(first, last)/Q/M=1 DAChannel
 			CHECK_EQUAL_VAR(V_max, amplitude)
 
 			// check that the level 3 pulse epoch is really only the pulse
-			if(level == 3)
-				WaveStats/R=(startT[i] * 1000 + EPOCHS_RESERVED_RANGE_MS, endT[i] * 1000 - EPOCHS_RESERVED_RANGE_MS)/Q/M=1 DAChannel
+			if(level == 3 && range > 2 * OTHER_EPOCHS_PRECISION)
+				WaveStats/R=(first + OTHER_EPOCHS_PRECISION, last - OTHER_EPOCHS_PRECISION)/Q/M=1 DAChannel
 				CHECK_EQUAL_VAR(V_min, amplitude)
 			endif
 		endif
 
 		// check baseline
-		if(strsearch(name, "Baseline", 0) > 0)
-			WaveStats/R=(startT[i] * 1000, endT[i] * 1000)/Q/M=1 DAChannel
+		if(strsearch(name, "Baseline", 0) > 0 && range > 2 * OTHER_EPOCHS_PRECISION)
+			WaveStats/R=(first, last)/Q/M=1 DAChannel
 			CHECK_EQUAL_VAR(V_min, 0)
 
 			// take something around the egdes due to decimation
-			WaveStats/R=(startT[i] * 1000 + EPOCHS_RESERVED_RANGE_MS, endT[i] * 1000 - EPOCHS_RESERVED_RANGE_MS)/Q/M=1 DAChannel
+			// offsets are in ms
+			WaveStats/R=(first + OTHER_EPOCHS_PRECISION, last - OTHER_EPOCHS_PRECISION)/Q/M=1 DAChannel
 			CHECK_EQUAL_VAR(V_max, 0)
 		endif
 	endfor

```

File: Packages/Testing-MIES/UTF_Epochs.ipf [hunk 1]
```diff
@@ -108,6 +108,7 @@ static Function TestEpochChannelTight(e)
 End
 
 static Constant OODDAQ_PRECISION = 0.001
+static Constant OTHER_EPOCHS_PRECISION = 0.05
```
Comment (MichaelHuth) [inline Packages/Testing-MIES/UTF_Epochs.ipf:111]: Just out of interest:
Are the 0.05 a result from the generation sample clock of the stimset or a convenient choice specifically for this test?

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
There is a typo "readd" in the commit message of
fdfc74b1faca482ec0baa562dd150349b6cf9bca

File: Packages/MIES/MIES_DataConfigurator.ipf [hunk 2]
```diff
@@ -2299,44 +2299,32 @@ static Function DC_AddEpochsFromStimSetNote(panelTitle, channel, stimset, stimse
...
-							subEpochEnd = limit(subEpochEnd, -Inf, epochEnd)
-							if(!pulseNr && subEpochBegin > epochBegin && subEpochbegin > stimsetBegin)
-								DC_AddEpoch(panelTitle, channel, epochBegin, subEpochBegin, EPOCH_BASELINE_REGION_KEY, 2, lowerlimit = stimsetBegin, upperlimit = stimsetEnd)
-							endif
-							epSubSubName = ReplaceNumberByKey("Pulse", epSubName, pulseNr, STIMSETKEYNAME_SEP, EPOCHNAME_SEP)
-							DC_AddEpoch(panelTitle, channel, subEpochBegin, subEpochEnd, epSubSubName, 2, lowerlimit = stimsetBegin, upperlimit = stimsetEnd)
+					endif
+
+					if(subEpochBegin >= epochEnd || subEpochEnd <= epochBegin)
+						DEBUGPRINT("Warning: sub epoch of flipped pulse starts after epoch end or ends before epoch start.")
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_DataConfigurator.ipf:2309]: Done.

File: Packages/Testing-MIES/UTF_Epochs.ipf (hunk 1 above)
Comment (t-b) [reply Packages/Testing-MIES/UTF_Epochs.ipf]: A mixture of playing around and deviation of stimset clock and DAQ clock.

**Comment by t-b** [comment]
`SC_SpikeControl` stimset gives

```
 from Calculation:
  '_free_'[0][0]= {100,2720.01}
'_free_'[0][1]= {519.99,619.99}
'_free_'[0][2]= {530,630}
  Warn: Returned pulse start times from Epochs but got none from Calculation. From Epochs:
  '_free_'[0][0]= {100,99.9999999999999,100,100,99.9999999999999,100,100,99.9999999999996,100,10.005}
'_free_'[0][1]= {1430.005,1530.005,1630.005,1730.005,1830.005,1930.005,2030.005,2130.005,2230.005,2330.005}
'_free_'[0][2]= {1440.005,1540.005,1640.005,1740.005,1840.005,1940.005,2040.005,2140.005,2240.005,2340.01}
  Warn: Returned pulse start times from Epochs but got none from Calculation. From Epochs:
  '_free_'[0][0]= {100,99.9999999999999,100,100,99.9999999999999,100,100,99.9999999999996,100,10.005}
'_free_'[0][1]= {1430.005,1530.005,1630.005,1730.005,1830.005,1930.005,2030.005,2130.005,2230.005,2330.005}
'_free_'[0][2]= {1440.005,1540.005,1640.005,1740.005,1840.005,1940.005,2040.005,2140.005,2240.005,2340.01}
```

investigate.

**Comment by t-b** [comment]
Should be good.

**Comment by t-b** [comment]
https://github.com/AllenInstitute/MIES/pull/811#issuecomment-778786360 Done and fixed.

**Comment by t-b** [comment]
I'm closing this one, this will not work as it is too closely intertwined with #788.

**Pull request closed without merging**